core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala

```diff
@@ -333,16 +333,17 @@ private[transaction] class TransactionMetadata(val transactionalId: String,
       (topicPartitions ++ addedTopicPartitions).toSet, newTxnStartTimestamp, updateTimestamp)
   }
 
-  def prepareAbortOrCommit(newState: TransactionState, clientTransactionVersion: TransactionVersion, nextProducerId: Long, updateTimestamp: Long): TxnTransitMetadata = {
+  def prepareAbortOrCommit(newState: TransactionState, clientTransactionVersion: TransactionVersion, nextProducerId: Long, updateTimestamp: Long, isEmptyAbort: Boolean): TxnTransitMetadata = {
     val (updatedProducerEpoch, updatedLastProducerEpoch) = if (clientTransactionVersion.supportsEpochBump()) {
       // We already ensured that we do not overflow here. MAX_SHORT is the highest possible value.
       ((producerEpoch + 1).toShort, producerEpoch)
     } else {
       (producerEpoch, lastProducerEpoch)
     }
 
+    val newTxnStartTimestamp = if (isEmptyAbort) updateTimestamp else txnStartTimestamp
```

Comment:
can we leave a comment explaining this? It might be confusing to those who don't know this code.